core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -283,25 +287,35 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
       try {
         transport = obtainActiveTransport();
       } catch (RuntimeException ex) {
-        stream = new NoopClientStream();
-        listener.closed(Status.INTERNAL.withDescription("Failed starting transport").withCause(ex),
-            new Metadata.Trailers());
+        closeCallPrematurely(listener,
+            Status.INTERNAL.withDescription("Failed starting transport").withCause(ex));
         return;
       }
       if (transport == null) {
-        stream = new NoopClientStream();
-        listener.closed(Status.CANCELLED.withDescription("Channel is shutdown"),
-            new Metadata.Trailers());
+        closeCallPrematurely(listener, Status.CANCELLED.withDescription("Channel is shutdown"));
         return;
       }
-      completeHeaders(headers);
+
+      // Convert the deadline to timeout. Timeout is more favorable than deadline on the wire
+      // because timeout tolerates the clock difference between machines.
+      long timeoutMicros = -1;
+      Long deadline = callOptions.getDeadline();
+      if (deadline != null) {
+        long timeoutMillis = deadline - System.currentTimeMillis();
```

Comment:
Hmm... So it isn't based on nanoTime... System.nanoTime() would be much better, as it is to be used for relative times. But this will bleed into our API I think.